# Семинар 14

$$\rho_{12;3} = \frac{\rho_{12} - \rho_{13}\rho_{23}}{\sqrt{(1 - \rho^2_{13})(1 - \rho^2_{23})}}$$
$\rho_{12;3} < \rho_{12}$ => связь между величинами частично обусловлена 3й величиной\
$\rho_{12;3} \to 0$ => связь между величинами полностью обусловлена 3й величиной\
$\rho_{12;3} \to 1$ => связь между величинами никак не обусловлена 3й величиной

### Задача 1

В некоторой области Англии исследовалось влияние погоды на урожай. Рассматривалось три показателя: урожай сена в центнерах на акр ($X_1$), весеннее количество осадков в дюймах ($X_2$) и накопленная за весну температура выше 42 градусов по Фаренгейту ($X_3$). По данным двадцатилетних наблюдений были вычислены реализации выборочных коэффициентов корреляции:
$$\hat \rho_{12} = 0.8; \hat \rho_{13} = −0.4; \hat \rho_{23} = −0.56$$

Проверьте гипотезу о некоррелированности показателей $X_1$ и $X_3$ против альтернативы
об их отрицательной коррелированности.\
Оценить частные коэффициенты корреляции $\hat \rho_{12;3}$, $\hat \rho_{13;2}$ и $\hat \rho_{23;1}$. 
Проверьте гипотезу о том, что частный коэффициент корреляции $\rho_{13;2}$ равен нулю. Прокомментировать полученный результат.

#### Проверка гипотезы о некоррелированности

Установим гипотезы:
- $H_0: \rho_{13} = 0$
- $H_a: \rho_{13} < 0$

Вычислим t-статистику:
$$t = \frac{\hat\rho_{13} \sqrt{n - 2}}{\sqrt{1 - \hat\rho^2_{13}}} = \frac{-0.4 \sqrt{18}}{\sqrt{1 - 0.16}} \approx -1.85$$
Найдем квантиль распределения Стьюдента на уровне значимости $0.05$ с $18$ степенями свободы - он равен примерно $-1.734$.\
Так как $t = -1.85 < -1.734$, мы отвергаем $H_0$ в пользу альтернативы $H_a$, что означает отрицательную корреляцию $X_1$ и $X_3$.

#### Рассчет частных коэффициентов корреляции

In [8]:
import numpy as np

def partial_corr(r12, r13, r23):
    """
    Calculate the partial correlation coefficient \hat{\rho}_{12;3} using the given correlation coefficients.

    Parameters:
    r12 : float
        Correlation coefficient between X1 and X2.
    r13 : float
        Correlation coefficient between X1 and X3.
    r23 : float
        Correlation coefficient between X2 and X3.

    Returns:
    float
        Partial correlation coefficient \hat{\rho}_{12;3}.
    """
    numerator = r12 - r13 * r23
    denominator = np.sqrt((1 - r13 ** 2) * (1 - r23 ** 2))
    r12_3 = numerator / denominator

    return r12_3

r12 = 0.8
r13 = -0.4
r23 = -0.56

rho_12_3 = partial_corr(r12, r13, r23)
rho_13_2 = partial_corr(r13, r12, r23)
rho_23_1 = partial_corr(r23, r12, r13)

rho_12_3, rho_13_2, rho_23_1

(0.7585674494892168, 0.09656090991705363, -0.43643578047198484)

$$\hat\rho_{12;3} = \frac{\hat\rho_{12} - \hat\rho_{13}\hat\rho_{23}}{\sqrt{(1 - \hat\rho^2_{13})(1 - \hat\rho^2_{23})}} = \frac{0.8 - (-0.4)(-0.56)}{\sqrt{(1 - (-0.4)^2)(1 - (-0.56)^2)}} \approx 0.759$$
$$\hat\rho_{13;2} = \frac{\hat\rho_{13} - \hat\rho_{12}\hat\rho_{23}}{\sqrt{(1 - \hat\rho^2_{12})(1 - \hat\rho^2_{23})}} \approx 0.097 \approx 0 \neq -0.4$$
$$\hat\rho_{23;1} = \frac{\hat\rho_{23} - \hat\rho_{12}\hat\rho_{13}}{\sqrt{(1 - \hat\rho^2_{12})(1 - \hat\rho^2_{13})}} \approx -0.436$$

#### Проверка гипотезы о частной корреляции

Установим гипотезы:
- $H_0: \rho_{13;2} = 0$
- $H_a: \rho_{13;2} \neq 0$

Рассчитаем t-статистику ($d$ - количество факторов):
$$t = \frac{\hat\rho_{13;2} \sqrt{n - 2 - d}}{\sqrt{1 - \hat\rho^2_{13;2}}} = \frac{0.097 \sqrt{20 - 2 - 1}}{\sqrt{1 - 0.097^2}} \approx 0.402$$
Найдем квантиль распределения Стьюдента на уровне значимости $0.05$ (для двустороннего теста) с $17$ степенями свободы - он равен примерно $2.110$.\
Так как $|t| = 0.402 < 2.110$, у нас нет достаточных оснований отвергнуть $H_0$, и мы принимаем $\rho_{13;2} = 0$

### Задача 2

Автосалон предоставил сведения о продажной цене ($X_1$), ширине ($X_2$), длине ($X_3$) и
массе ($X_4$) автомобиля. За последний месяц было продано 34 автомобиля. На основании
этих данных вычислены выборочные коэффициенты корреляции: 
$$\hat \rho_{12} = 0.33, \hat \rho_{13} = 0.16, \hat \rho_{14} = 0.53, \hat \rho_{23} = 0.71, \hat \rho_{24} = 0.72, \hat \rho_{34} = 0.63$$
Оценить множественный коэффициент корреляции $R_{1(2,3,4)}$ между продажной ценой
автомобиля и совокупностью его трех технических характеристик, описывающих
длину, высоту и массу. Проверить гипотезу о том, что $R_{1(2,3,4)} = 0$, предполагая, что
данные имеют гауссовское распределение. Прокомментировать полученный результат.

Множественный коэф. корреляции:
$$\mathbb{R}_{1(2,3,4)} = \sqrt{1 - \frac{\det \mathbb{R}}{\mathbb{R}_{11}}}$$

Матрица корреляции:
$$\mathbb{R} = 
\begin{pmatrix}
    1 && \hat\rho_{12} && \hat\rho_{13} && \hat\rho_{14}\\
    \hat\rho_{12} && 1 && \hat\rho_{23} && \hat\rho_{24}\\
    \hat\rho_{13} && \hat\rho_{23} && 1 && \hat\rho_{34}\\
    \hat\rho_{14} && \hat\rho_{24} && \hat\rho_{34} && 1\\
\end{pmatrix} = 
\begin{pmatrix}
    1 && 0.33 && 0.16 && 0.53\\
    0.33 && 1 && 0.71 && 0.72\\
    0.16 && 0.71 && 1 && 0.63\\
    0.53 && 0.72 && 0.63 && 1\\
\end{pmatrix}
$$

Дополнительная матрица:
$$\mathbb{R}_{11} =
\begin{pmatrix}
    1 && 0.71 && 0.72\\
    0.71 && 1 && 0.63\\
    0.72 && 0.63 && 1\\
\end{pmatrix}
$$

In [17]:
R = np.array(
    [[1, 0.33, 0.16, 0.53],
    [0.33, 1, 0.71, 0.72],
    [0.16, 0.71, 1, 0.63],
    [0.53, 0.72, 0.63, 1]])

R11 = np.array(
    [[1, 0.71, 0.72],
    [0.71, 1, 0.63],
    [0.72, 0.63, 1]])

det_R = np.linalg.det(R)
det_R11 = np.linalg.det(R11)

det_R, det_R11

(0.15002191999999998, 0.224712)

Определитель матрицы корреляции:
$$\det \mathbb{R} \approx 0.15$$

Алгебраическое дополнение:
$$\mathbb{R} \approx 0.025$$

In [20]:
res = np.sqrt(1 - (det_R / det_R11))
res

0.5765252459788318

$$\mathbb{R}_{1(2,3,4)} = \sqrt{1 - \frac{0.15}{0.225}} \approx 0.5765$$

Проверим гипотезу:
- $H_0: R_{1(2,3,4)} = 0$
- $H_a: R_{1(2,3,4)} \neq 0$

In [23]:
res_squared = 1 - (det_R / det_R11)
res_squared

0.33238135925095247

In [24]:
f_statistics = (1/3 * res_squared) / (1/30 * (1 - res_squared))
f_statistics

4.978611125627511

Критерий Фишера:
$$F = \frac{\frac{1}{l - 1} R^2_{1(2,3,4)}}{\frac{1}{n - l} (1 - R^2_{1(2,3,4)})} = \frac{\frac{1}{4 - 1} \cdot 0.332}{\frac{1}{34 - 4} (1 - 0.332)} \approx 4.979 \sim F(l - 1; n - 1)$$

In [26]:
import scipy.stats as stats

alpha = 0.05
df1 = 3
df2 = 30

f_crit = stats.f.ppf(1 - alpha, df1, df2)
f_crit

2.9222771906450378

$$F_{crit} = F(3, 30) \approx 2.92$$
$$F > F_{crit}$$

На уровне значимости $\alpha = 0.05$ мы отвергаем нулевую гипотезу о равенстве множественного коэффициента корреляции нулю. Таким образом, существует статистически значимая связь между продажной ценой автомобиля и его техническими характеристиками.

### Задача 3

Три квалифицированных эксперта $A$, $B$ и $C$ проранжировали в порядке предпочтения
семь представленных бизнес-проектов.\
Результаты представлены в таблице.

|   | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
|---|---|---|---|---|---|---|---|
| A | 1 | 4 | 2 | 5 | 3 | 7 | 6 |
| B | 2 | 1 | 3 | 4 | 5 | 6 | 7 |
| C | 2 | 1 | 4 | 5 | 3 | 7 | 6 |

Можно ли считать, что данная экспертная группа обладает общей системой
предпочтений?

Воспользуемся коэффициентом конкордации Кендалла:
$$W = \frac{12S}{m^2(n^3 - n)}$$

- $m = 3$ - количество экспертов
- $n = 7$ - количество объектов
- $S$ - сумма квадратов отклонений рангов объектов от их среднего ранга

Рассчитаем сумму рангов для проектов:

|   | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
|-----------|---|---|---|---|---|---|---|
| A         | 1 | 4 | 2 | 5 | 3 | 7 | 6 |
| B         | 2 | 1 | 3 | 4 | 5 | 6 | 7 |
| C         | 2 | 1 | 4 | 5 | 3 | 7 | 6 |
| A + B + C | 5 | 6 | 9 | 14 | 11 | 20 | 19 |

Найдем средний ранг: 
$$\bar{R} = \frac{m(n+1)}{2} = 12$$

Вычислим сумму квадратов отклонений рангов объектов от их среднего ранга:
$$S = \sum_{i=1}^n(R_i - \bar{R})^2 = (5−12)^2 + (6−12)^2 + (9−12)^2 + (14−12)^2 + (11−12)^2 + (20−12)^2 + (19−12)^2 = 212$$

Рассчитаем $W$:
$$W = \frac{12S}{m^2(n^3 - n)} = \frac{12 \cdot 212}{3^2(7^3 - 7)} = \frac{53}{63} \approx 0.841$$

In [36]:
alpha = 0.05
df = 6

chi_crit = stats.chi2.ppf(1 - alpha, df)
chi_crit

12.591587243743977

Оценим значение с помощью хи-квадрат статистики:
$$\chi^2 = m(n - 1)W = 3(7 - 1) \cdot \frac{53}{63} = \frac{106}{7} \approx 15.14$$

Критическое значение хи-квадрат статистики на уровне значимости $\alpha = 0.05$ с $n - 1 = 6$ степенями свободы будет равно $12.592$. 

Так как $\chi^2 > \chi^2_{crit}$, мы отвергаем $H_0$ в пользу альтернативы $H_a$, что означает: данная экспертная группа обладает общей системой предпочтений.